Add scheduled cleanup for remote posts (ap_post)

pfefferle opened the pull request:
## Proposed changes:
* Add scheduled purge functionality for remote ActivityPub posts (`ap_post`) similar to existing outbox and inbox cleanup
* Refactor purge logic by moving it from Scheduler into the respective collection classes (Outbox, Inbox, Posts)

### Key features:
- Preserves posts that have comments (likes, reposts, quotes) from local users as these indicate meaningful interactions
- Uses same threshold (200 posts) and default retention (180 days) as inbox
- Daily cron schedule (`activitypub_ap_post_purge`)
- Option `activitypub_ap_post_purge_days` not user-facing yet (no settings UI)
- Added to Site Health info for debugging

### Other information:

- [x] Have you written new tests for your changes, if applicable?

## Testing instructions:

* Run the new tests: `npm run env-test -- --filter="Test_Scheduler::test_purge_ap_posts"`
* Verify all 4 new tests pass:
  - `test_purge_ap_posts_more_than_200_posts` - verifies deletion when threshold exceeded
  - `test_purge_ap_posts_200_or_fewer_posts` - verifies no deletion below threshold
  - `test_purge_ap_posts_preserves_posts_with_comments` - verifies posts with interactions are kept
  - `test_purge_ap_posts_with_different_purge_days` - verifies configurable retention period
* Run all purge tests to ensure refactoring didn't break existing functionality: `npm run env-test -- --filter="Test_Scheduler::test_purge"`

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [x] Minor

#### Type

-   [x] Added - for new features

#### Message

Add scheduled cleanup for remote posts, preserving posts with local user interactions.

</details>

pfefferle committed:
Add scheduled cleanup for remote posts (ap_post)

Add purge functionality for remote ActivityPub posts similar to
existing outbox and inbox cleanup. Key features:

- Preserves posts that have comments (likes, reposts, quotes) from
  local users as these indicate meaningful interactions
- Uses same threshold (200 posts) and default retention (180 days)
  as inbox
- Option not user-facing yet (no settings UI)

Also refactors purge logic by moving it from Scheduler into the
respective collection classes (Outbox, Inbox, Posts), making the
code more maintainable and allowing direct use of purge methods.

matticbot committed:
Add changelog

pfefferle committed:
Add unit tests for collection purge methods

Add comprehensive tests for the purge() methods in Outbox, Inbox,
and Posts collection classes:

- Test threshold behavior (20 for Outbox, 200 for Inbox/Posts)
- Test retention period configuration
- Test preservation of Follow activities in Outbox
- Test preservation of posts with comments in Posts collection
- Verify return count of deleted items

obenland commented on includes/wp-admin/class-health-check.php:
```diff
@@ -342,6 +342,12 @@ public static function debug_information( $info ) {
 
+		$info['activitypub']['fields']['activitypub_ap_post_purge_days'] = array(
+			'label'   => \__( 'Remote Posts Retention Period', 'activitypub' ),
+			'value'   => \esc_attr( (int) \get_option( 'activitypub_ap_post_purge_days', 180 ) ),
```
I'd keep it much lower, like a week or a month, if my test sites are any indicator.

<img width="210" height="86" alt="Image" src="https://github.com/user-attachments/assets/e964718a-5c76-4bbf-88ef-d2a31a4fc3e8" />

obenland commented on includes/wp-admin/class-settings.php:
```diff
@@ -186,6 +186,17 @@ public static function register_settings() {
+				'type'         => 'integer',
+				'description'  => \__( 'Number of days to keep remote posts.', 'activitypub' ),
+				'default'      => 180,
+				'show_in_rest' => false,
```
```suggestion
				'description'  => 'Number of days to keep remote posts.',
				'default'      => 30,
```
I'd keep it much lower, like a week or a month. Is `show_in_rest` not false by default?

obenland commented on includes/class-scheduler.php:
```diff
@@ -315,66 +323,24 @@ public static function reprocess_outbox() {
-			\wp_delete_post( $post_id, true );
-		}
+		$days = (int) get_option( 'activitypub_outbox_purge_days', 180 );
+		Outbox::purge( $days );
```
At this point you're just two steps away from replacing `purge_outbox` with `Outbox::purge`. It would just need to move the `get_option()` call and update the hook callback in line 63 (and following)

pfefferle replied:
This was on purpose. I wanted to keep them separate, to maybe have CLI commands that you can call to purge by hand.

obenland commented on includes/class-scheduler.php:
```diff
@@ -452,6 +418,20 @@ public static function handle_inbox_purge_days_update( $old_value, $value ) {
+	 * @param int $old_value The old value.
+	 * @param int $value     The new value.
+	 */
+	public static function handle_ap_post_purge_days_update( $old_value, $value ) {
```
```suggestion
	public static function update_ap_post_purge_schedule( $old_value, $value ) {
```

obenland commented on includes/collection/class-posts.php:
```diff
@@ -452,4 +452,50 @@ public static function remove_recipient( $post_id, $user_id ) {
+
+		$deleted = 0;
+		foreach ( $post_ids as $post_id ) {
+			// Preserve posts with comments (includes likes, reposts, quotes).
```
What will happen with them?

pfefferle replied:
I thought users might want to keep posts they interacted with!?

pfefferle replied:
especially if we only keep posts for 30 days!?

obenland replied:
This would keep them forever. I guess we could do that.

I just assumed that we'd remote request them just-in-time as needed, like when someone would look at all their likes (if and when we'd offer that) or something. Not sure.

pfefferle replied:
the problem is, that bringing the post up again after a new comment with all the comment history is not possible for now and might cause quite some time to implement.

obenland replied:
This is probably good enough then for now.

pfefferle committed:
Merge branch 'trunk' into add/ap-post-purge

pfefferle committed:
Merge branch 'trunk' into add/ap-post-purge

pfefferle committed:
Merge branch 'trunk' into add/ap-post-purge

Menrath commented on includes/collection/class-inbox.php:
```diff
@@ -464,4 +464,42 @@ public static function deduplicate( $guid ) {
+		);
+
+		$deleted = 0;
+		foreach ( $post_ids as $post_id ) {
```
Could you maybe add a filter here, or for the query above? For events it may be crucial not to delete them before the `endTime`.

pfefferle replied:
good point 🤔

pfefferle committed:
Address PR feedback for remote posts purge feature

- Change default retention for remote posts from 180 to 30 days
- Remove translation wrapper from setting description
- Rename handler methods from handle_*_purge_days_update to update_*_purge_schedule
- Update tests to explicitly set option values

pfefferle committed:
Fix ap_post purge to only preserve posts with local user comments

pfefferle committed:
Add filter to skip remote post deletion during purge

pfefferle committed:
Update tests to use local user comments for purge preservation

obenland commented on includes/collection/class-posts.php:
```diff
@@ -452,4 +452,73 @@ public static function remove_recipient( $post_id, $user_id ) {
+					$post_id
+				)
+			);
+			if ( $local_comments > 0 ) {
```
```suggestion
			$has_local_comments = (bool) $wpdb->get_var(
				$wpdb->prepare(
					"SELECT 1 FROM $wpdb->comments WHERE comment_post_ID = %d AND user_id > 0 LIMIT 1",
					$post_id
				)
			);
			if ( $has_local_comments ) {
```
Micro optimization as we're not interested in the actual IDs, just wether they exist.

obenland commented on includes/class-scheduler.php:
```diff
@@ -315,66 +323,24 @@ public static function reprocess_outbox() {
-		foreach ( $post_ids as $post_id ) {
-			\wp_delete_post( $post_id, true );
-		}
+		$days = (int) get_option( 'activitypub_outbox_purge_days', 180 );
```
```suggestion
		$days = (int) \get_option( 'activitypub_outbox_purge_days', 180 );
```

obenland commented on includes/class-scheduler.php:
```diff
@@ -315,66 +323,24 @@ public static function reprocess_outbox() {
-				),
-			)
-		);
+		$days = (int) get_option( 'activitypub_inbox_purge_days', 180 );
```
```suggestion
		$days = (int) \get_option( 'activitypub_inbox_purge_days', 180 );
```

obenland commented on includes/class-scheduler.php:
```diff
@@ -315,66 +323,24 @@ public static function reprocess_outbox() {
+	 * Purge remote posts based on a schedule.
+	 */
+	public static function purge_ap_posts() {
+		$days = (int) get_option( 'activitypub_ap_post_purge_days', 30 );
```
```suggestion
		$days = (int) \get_option( 'activitypub_ap_post_purge_days', 30 );
```

obenland commented on tests/phpunit/tests/includes/collection/class-test-inbox.php:
```diff
@@ -1102,4 +1102,154 @@ public function test_add_like_activity_with_trailing_slash() {
+			array(
+				'post_type'   => Inbox::POST_TYPE,
+				'post_status' => 'publish',
+				'post_date'   => gmdate( 'Y-m-d H:i:s', strtotime( '-7 months' ) ),
```
Doesn't look like much of the tests code added here has backslashes

pfefferle committed:
class-posts.php aktualisieren

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
class-scheduler.php aktualisieren

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
class-scheduler.php aktualisieren

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
class-scheduler.php aktualisieren

Co-authored-by: Konstantin Obenland <obenland@gmx.de>

pfefferle committed:
Backslashify global PHP/WordPress functions in purge tests

Add backslash prefix to global functions in namespaced test files
for improved performance and consistency with coding standards.

pfefferle committed:
Merge branch 'trunk' into add/ap-post-purge
